### PR TITLE
Improve loading of update steps

### PR DESCRIPTION
**Pull request by solidgoldpig**
- Enable multiple steps to be loaded per journey
- Enable different role per journey
- Use key for urls and locales lookup rather than brittle url munging
- Process all update link data up front in move view
- Add ability to moveSummaryList component to output action links
- Enable link to update move date
- Suppress updating of to location route until redirect endpoint in place
- Ensure e2e tests actually change gender

Move_details_for_MCDERMOTT__JAMES

### Checklist

- [x] If adding new environment variables, they have been documented in the [README](ministryofjustice/hmpps-book-secure-move-frontend/blob/master/README.md)
- [x] No new [Personally Identifiable Information (PII)](https://support.google.com/analytics/answer/6366371?hl=en) is being sent via analytics

### CODE REVIEW

File: app/move/controllers/update/base.js [hunk 1]
```diff
@@ -76,7 +76,7 @@ class UpdateBaseController extends CreateBaseController {
 
       try {
         if (req.initialStep) {
-          values = this.getUpdateValues(req)
+          values = this.getUpdateValues(req, res)
```
Comment (solidgoldpig) [inline app/move/controllers/update/base.js:79]: Missed this in previous PR

File: app/move/views/_includes/assessment.njk [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {% for assessmentCategory in assessment %}
-  <section class="app-!-position-relative govuk-!-margin-top-7" data-app-section="{{ assessmentCategory.key }}">
+  <section class="app-!-position-relative govuk-!-margin-top-7">
```
Comment (solidgoldpig) [inline app/move/views/_includes/assessment.njk:2]: Not needed any more

File: test/e2e/pages/create-move.js [hunk 3]
```diff
@@ -187,7 +187,7 @@ class CreateMovePage extends Page {
       },
       gender: {
         selector: this.fields.gender,
-        value: faker.random.arrayElement(['Male', 'Female']),
+        value: person.gender,
```
Comment (solidgoldpig) [inline test/e2e/pages/create-move.js:190]: Random selection now handled by `generatePerson`

**Comment by jason-sims** [comment]
tag when you have the build that you want me to review. cannot test the active ff for police user with this deployment. May be easier to review together. Cheers.

**Comment by solidgoldpig** [comment]
> tag when you have the build that you want me to review. cannot test the active ff for police user with this deployment. May be easier to review together. Cheers.

Probably better to wait for the confirmation messages to drop first (which will be along soon - just writing the tests)

File: app/move/controllers/view/view.update.links.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
+const getUpdateLinks = (updateSteps, t, urls) => {
+  const updateLinks = {}
+  updateSteps.forEach(updateJourney => {
+    const category = updateJourney.key
+    if (!urls[category]) {
+      return
+    }
+    const categoryText = t('moves::update_link.categories.' + category)
```
Comment (barbaracassani) [inline app/move/controllers/view/view.update.links.js:8]: I was under the impression that we enforced the use of html template strings for string concatenation, but I might be wrong.

File: test/e2e/_move.js [hunk 5]
```diff
@@ -132,13 +132,7 @@ export async function checkNoUpdateLink(page) {
...
 }
 
-const updatePages = [
-  'personal_details',
-  'risk',
-  'health',
-  'court',
-  'move_details',
-]
+const updatePages = ['personal_details', 'risk', 'health', 'court', 'date']
```
Comment (barbaracassani) [inline test/e2e/_move.js:135]: The dramatic changes in formatting when you rename / remove an element are always funny.

File: app/move/controllers/view/view.update.links.js (hunk 4 above)
Comment (teneightfive) [reply app/move/controllers/view/view.update.links.js:8]: I think its the preferred pattern but maybe we're missing an eslint rule for it...

File: app/move/steps/update.js [hunk 6]
```diff
@@ -1,44 +1,88 @@
...
+  buttonText: 'actions::save_and_continue',
 }
 
-const updateSteps = Object.keys(createSteps)
-  .filter(key => updateControllers[key])
-  .map(key => ({ [key]: provideStepProps(key, createSteps[key]) }))
+const updateSteps = [
+  {
+    key: 'personal_details',
+    role: 'move:update',
```
Comment (teneightfive) [inline app/move/steps/update.js:21]: I wonder if this is the wrong term. We've tried really hard to not have "roles" within the app.

When a user signs in, with given "roles", we assign permissions, so maybe if that's what we're trying to do here we should use `permissions` as an array for consistency?

File: common/components/meta-list/template.njk [hunk 7]
```diff
@@ -1,7 +1,7 @@
 <dl class="app-meta-list {% if params.classes -%} {{ params.classes }}{%- endif %}">
   {% for item in params.items %}
     {% if item.value.text or item.value.html %}
-      <div class="app-meta-list__item">
+      <div class="app-meta-list__item app-!-position-relative">
```
Comment (teneightfive) [inline common/components/meta-list/template.njk:4]: Could this be added using `params.classes` rather than this class applied and style being applied to all areas where this component is used.

File: common/components/meta-list/template.njk [hunk 8]
```diff
@@ -11,6 +11,15 @@
         <dd class="app-meta-list__value">
           {{ item.value.html | safe if item.value.html else item.value.text }}
         </dd>
+
+        {% set updateLink = item.link %}
+        {% if updateLink %}
```
Comment (teneightfive) [inline common/components/meta-list/template.njk:16]: Also, the addition of this no longer feels content agnostic. The naming of the variables is binding it specifically to the link saying "Update" or "Change", whereas in the terms of a component, it should have more flexibility.

I think it would be cleaner for this component to support a similar pattern to the [GOV.UK Summary list component](https://design-system.service.gov.uk/components/summary-list/) where the component can have "actions" and the presenter sends in the text and href constructed from the URLs object.

Comment (teneightfive) [reply common/components/meta-list/template.njk:16]: I think this pattern of "actions" would be nice to use actually:

```
...
actions: {
  items: [
    {
      href: "#",
      text: "Change",
      visuallyHiddenText: "name"
    }
  ]
}
...
```

File: app/move/controllers/view/view.update.links.js (hunk 4 above)
Comment (solidgoldpig) [reply app/move/controllers/view/view.update.links.js:8]: We don't enforce it - but it is better

File: app/move/steps/update.js (hunk 6 above)
Comment (solidgoldpig) [reply app/move/steps/update.js:21]: `permission` fine. Why an array though? This gets passed to `protectRoute` which takes a string

Comment (teneightfive) [reply app/move/steps/update.js:21]: Ah yes. That middleware only support a string. And if we need to protect based on multiple permissions we have used that middleware twice.

Maybe you need to cater for that here though. A user may require `move:update` as well as `move:update:documents` for a particular step. Passing just `move:update:documents` wouldn't feel enough as we would require that user to have the permission to update first as well as update at a granular level.

Comment (teneightfive) [reply app/move/steps/update.js:21]: Is this to support STC/SCH where only they can edit/upload documents?

At the moment, that logic is actually tied to the `from_location` rather than a specific permission because of the way it was originally required. We may want the ability to allow other establishments to upload documents, regardless of that user's roles/permissions.

If it's just to support documents at this stage, this may not be the best place for that logic if we need to keep it tied to `from_location`. That logic could live in the [documents controller](https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/blob/95be55df32857b54a6a3193cba8c469b524ebf20/app/move/controllers/create/document.js#L16) instead.

File: common/components/meta-list/template.njk (hunk 8 above)
Comment (solidgoldpig) [reply common/components/meta-list/template.njk:16]: tldr; the meta-list component is already a specialist component with hard-coded fields in it.

That feels like premature optimisation? Do we know that the metalist requires multiple actions?  The following would do what you suggest, but it can only do it by subsuming information about classes and attributes which to me seems even yuckier. 

`common/components/meta-list/template.njk`

```njk
{%- macro _actionLink(action) %}
  <a class="govuk-link {%- if action.classes %} {{ action.classes }}{% endif %}" href="{{ action.href }}" {%- for attribute, value in action.attributes %} {{attribute}}="{{value}}"{% endfor %}>
    {{ action.html | safe if action.html else action.text }}
    {%- if action.visuallyHiddenText -%}
      <span class="govuk-visually-hidden"> {{ action.visuallyHiddenText }}</span>
    {% endif -%}
  </a>
{% endmacro -%}

<dl class="app-meta-list {% if params.classes -%} {{ params.classes }}{%- endif %}">
  {% for item in params.items %}
    {% if item.value.text or item.value.html %}
      <div class="app-meta-list__item {% if item.classes -%} {{ item.classes }}{%- endif %}">
        {% if item.key.html or item.key.text %}
          <dt class="app-meta-list__key">
            {{ item.key.html | safe if item.key.html else item.key.text }}
          </dt>
        {% endif %}

        <dd class="app-meta-list__value">
          {{ item.value.html | safe if item.value.html else item.value.text }}
        </dd>

        {% if item.actions.items.length %}
          <dd class="app-meta-list__actions {%- if item.actions.classes %} {{ item.actions.classes }}{% endif %}">
            {% if item.actions.items.length == 1 %}
              {{ _actionLink(item.actions.items[0]) | indent(12) | trim }}
            {% else %}
              <ul class="app-meta-list__actions-list">
                {% for action in item.actions.items %}
                  <li class="app-meta-list__actions-list-item">
                    {{ _actionLink(action) | indent(18) | trim }}
                  </li>
                {% endfor %}
              </ul>
            {% endif %}
          </dd>
        {% endif %}
      </div>
    {% endif %}
  {% endfor %}
</dl>
```

`common/presenters/move-to-meta-list-component.js`

```js
...
    additional_information: additionalInfo,
  } = {},
  additionalProps = {}
) {
  const destination = get(toLocation, 'title', 'Unknown')
  const destinationLabel =
    moveType === 'prison_recall'
      ? i18n.t('fields::move_type.items.prison_recall.label')
      : destination
  const additionalInformation = additionalInfo ? ` — ${additionalInfo}` : ''
  const items = [
    {
      key: {
        text: i18n.t('fields::from_location.short_label'),
      },
      value: {
        text: get(fromLocation, 'title'),
      },
    },
    {
      ...additionalProps.move,
      key: {
        text: i18n.t('fields::move_type.short_label'),
      },
      value: {
        text: destinationLabel + additionalInformation,
      },
    },
    {
      ...additionalProps.date,
      key: {
        text: i18n.t('fields::date_type.label'),
      },
      value: {
        text: setDateToDisplay({ date, dateFrom, dateTo }),
      },
    },
...
```
`app/move/controllers/view.js`

```js
...
  const updateLinks = getUpdateLinks(updateSteps, req.t, updateUrls)

  const urls = {
    update: updateUrls,
  }

  const getSummaryItemProps = actions => {
    if (!actions) {
      return
    }
    return {
      classes: 'app-!-position-relative',
      actions: {
        classes: 'app-!-position-top-right govuk-!-font-size-16',
        items: [actions],
      },
    }
  }

  const metaListProps = {
    date: getSummaryItemProps(updateLinks.date),
    move: getSummaryItemProps(updateLinks.move),
  }

  const locals = {
    moveSummary: presenters.moveToMetaListComponent(move, metaListProps),
    personalDetailsSummary: presenters.personToSummaryListComponent(person),
...
```

`app/move/controllers/view/view.update.links.js`
```js
...
      updateLinks[category] = {
        category,
        attributes: {
          'data-update-link': category,
        },
        href: urls[category],
        html: t('moves::update_link.link_text', {
          context: 'with_category',
          category: categoryText,
        }),
      }
...
```

Can we leave this for more discussion and not block getting this in for now?

Comment (teneightfive) [reply common/components/meta-list/template.njk:16]: I hear you about we might not need multiple at this stage so that may be overkill. My point was more around this component being more agnostic to the content of the links you're currently adding.

Like with the GOV.UK example, I think the difference between a component and using elements within a template or specific view is that a component should be more generic in its use. The name of things suggests that this would only ever be used to add an "Update" link, but in reality the component doesn't care what the link says.

I also understand we don't want to hold changes up, but I think changes to components just need to be thought of a bit more to not tie them down to single use, otherwise they lost the ability to be as easily re-used.

Comment (teneightfive) [reply common/components/meta-list/template.njk:16]: The change needed could also be handled by using `item.value.html` instead of `item.value.text` and adding the link in the presenter.

And wouldn't require any updates to the component.

File: app/move/steps/update.js (hunk 6 above)
Comment (solidgoldpig) [reply app/move/steps/update.js]: It also allows canAccess to be called within the controller using the value from the config

File: common/components/meta-list/template.njk (hunk 8 above)
Comment (solidgoldpig) [reply common/components/meta-list/template.njk]: Updated as discussed with meta-list component gaining action and not requiring app-!-style overrides

File: common/components/meta-list/template.njk (hunk 7 above)
Comment (solidgoldpig) [reply common/components/meta-list/template.njk]: Made redundant by updating the component styling

File: app/move/controllers/view/view.update.links.js [hunk 9]
```diff
@@ -0,0 +1,26 @@
+const getUpdateLinks = (updateSteps, t, urls) => {
```
Comment (teneightfive) [inline app/move/controllers/view/view.update.links.js:1]: Rather than having to pass in `t` you can directly import the [i18n module](https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/blob/5f692a046faea2298687c7a23bc4882734c7a60b/config/i18n.js) and call `i81n.t` on that.

Here's [an example](https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/blob/5f692a046faea2298687c7a23bc4882734c7a60b/common/presenters/move-to-card-component.js#L1).

File: app/move/controllers/view/view.update.urls.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+const getUpdateUrls = (updateSteps, moveId, canAccess) => {
```
Comment (teneightfive) [inline app/move/controllers/view/view.update.urls.js:1]: I think rather than passing in the `canAccess` function which is a view helper it would be cleaner to pass in the current user's permissions and use the [permissions module](https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/blob/5f692a046faea2298687c7a23bc4882734c7a60b/common/middleware/permissions.js#L3) to do this check.

File: app/move/controllers/view.js [hunk 11]
```diff
@@ -16,22 +19,24 @@ module.exports = function view(req, res) {
...
-    urls.update[key] = `/move/${move.id}/edit/${url}`
-    // some assessment categories aren't suffixed with _information
-    if (key.endsWith('_information')) {
-      const keyWithoutInformation = key.replace(/_information$/, '')
-      urls.update[keyWithoutInformation] = urls.update[key]
+
+  const metaListActions = {}
+  Object.keys(updateLinks).forEach(key => {
+    metaListActions[key] = {
+      classes: 'app-meta-list__action--sidebar',
```
Comment (teneightfive) [inline app/move/controllers/view.js:33]: It feels like all of this should live in the presenter. It's purely related to how we display the actions when they're in that component so feels better to contain them there rather than in this controller.

This controller is starting to contain a lot of logic and we've previously tried (where possible) to keep thin controllers and extract any complicated logic to smaller modules.

File: common/components/meta-list/template.njk [hunk 12]
```diff
@@ -11,6 +11,14 @@
         <dd class="app-meta-list__value">
           {{ item.value.html | safe if item.value.html else item.value.text }}
         </dd>
+
+        {% if item.action.html or item.action.text %}
+          <dd class="app-meta-list__action {% if item.action.classes -%} {{ item.action.classes }}{%- endif %}">
```
Comment (teneightfive) [inline common/components/meta-list/template.njk:16]: Thanks for these changes. This feels much cleaner from a component perspective and removes the tie to the content of the element.

However, it would be good if this could be split into a separate commit with a little explanation why we changed it (maybe once everything has been rebased).

If you look back on that component with `git blame` embedded in the rest of this PR it will be a bit noisy.

File: app/move/controllers/view/view.update.links.js (hunk 9 above)
Comment (solidgoldpig) [reply app/move/controllers/view/view.update.links.js:1]: Sure, but there is a slight werdness there. I had done that and got bitten by using 

```
const { t } = require('../../../../config/i18n')
```

and getting

```
error: Cannot read property 'translator' of undefined {"stack":"TypeError: Cannot read property 'translator' of undefined\n    at t
```

whereas _obviously_ I should have been doing ;)

```js
const i18n = require('../../../../config/i18n')

module.exports = () => {
...
   const fooStr = i18n.t('foo')
}
```

Maybe it's worth waiting for i18n to actually be properly instantiated before it gets exported. Not a biggy, but principle of least surprise...

File: app/move/controllers/view/view.update.urls.js (hunk 10 above)
Comment (solidgoldpig) [reply app/move/controllers/view/view.update.urls.js]: Done

File: app/move/controllers/view.js (hunk 11 above)
Comment (solidgoldpig) [reply app/move/controllers/view.js]: Done

File: common/components/meta-list/template.njk (hunk 12 above)
Comment (solidgoldpig) [reply common/components/meta-list/template.njk:16]: Rejigged

File: app/move/controllers/view/view.update.links.js (hunk 9 above)
Comment (teneightfive) [reply app/move/controllers/view/view.update.links.js]: I also find the second approach is much testable as you can use sinon to stub the `t` method. Whereas if importing using destructuring you would have to use something lik proxyquire to stub and it gets messy...
